### PR TITLE
Advanced Building Blocks - Enumerables Project

### PR DESCRIPTION
**Pull request by RyelBanfield**
This PR contains the entire project built on a feature branch. It features custom built enumerable methods.

### CODE REVIEW

**Review by BrianSammit — CHANGES_REQUESTED** [review]
CHANGES REQUIRED :recycle: :red_circle: 
-----------------
Hello team!! :wave: you are doing an amazing job :100: 
However, there are changes to make :hammer_and_wrench: see them below :eyes: :point_down: 
--------------

- [x] On the method `#my_each`, I got this error (undefined local variable or method `i`), you are not defining the `i ` variable on this method. 
- [x] Change your string class method `length` to `size`, `length` gives troubles with ranges `5...50`.

with the changes above :point_up_2: check the following ones.
- [x] #my_each returns an Enumerator if no block is given. 
- [x] #my_each when a block is given, when self is an array returns the array itself after calling the given block once for each element in self. Instead is returning `nil`.
- [x] #my_each when a block is given when self is a range returns the range itself after calling the given block once for each element in self. Instead is returning `nil`.
- [ ] #my_each when a block is given when self is a hash behaves like #each calls the given block once for each element in self.
- [ ] #my_each_with_index returns an enumerator if no block is given.
- [ ] #my_each_with_index when a block is given when self is an array returns the array itself after calling the given block once for each element in self. Instead is returning `nil`.
- [ ] #my_each_with_index when a block is given when self is a range returns the range itself after calling the given block once for each element in self. Undefined method `[]`.
------------------
Make these required changes :point_up_2: and send it for another review :muscle: 
HAPPY CODDING!!

AngelicConcreteHypsilophodon-max-1mb

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Mert Gündüz & Ryel Banfield** 
Good job with your first submission for this project :muscle:  :muscle:
You have made an important progress :clap: :clap: :clap:  
We still need to make some improvement before you can go ahead:

- [x] Add a screenshot in your README file
- [x] Don't overuse rubocop disable
- [x] #my_each when a block is given when self is a range returns the range itself after calling the given block once for each element in 
self
        `[1,2,3].my_each(&proc{|x| x>2}) == [1,2,3]`
- [x] #my_all? returns false if the block returns false or nil::range
    `(1..3).all?(&proc{|x| x%2==0}) == (1..3).my_all?(&proc{|x| x%2==0})`
- [x] #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern      
      `[5,5,[5]].all?(5) == [5,5,[5]].my_all?(5)`
- [x] #my_any? returns false if the block ever returns a value false or nil::range
    `(1..3).any?(&proc{|x| x==0}) == (1..3).my_any?(&proc{|x| x==0})`
- [x] #my_any? when no block or argument is given returns false if at least one of the collection is not true
      `[false, 0].any? == [false, 0].my_any?`
- [x] #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern
      `['cat', 'dog','car'].any?('cat') == ['cat', 'dog','car'].my_any?('cat')`
- [x] #my_none? returns false if the block ever returns true for all elements::range
    `(1..3).none?(&proc{|num| num%2==0}) == (1..3).my_none?(&proc{|num| num%2==0})`
- [ ] #my_none? when a Regex is passed as an argument returns false if any of the collection
matches the Regex
      `['dog', 'cat'].none?(/d/) == ['dog', 'cat'].my_none?(/d/)`
- [ ] #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern       
      `[5,'dog','car'].none?(5) == [5,'dog','car'].my_none?(5)`
- [ ] #my_count returns the number of items in enum through enumeration::range
    `(1..3).count == (1..3).my_count`
- [ ] #my_map returns a new array with the results of running a given block once for every element in enum::range
    `(1..3).map(&proc{|num| num+1}) == (1..3).my_map(&proc{|num| num+1})`
- [ ] #my_inject searches for the longest word in an array of strings
    `['to','four','tri'].my_inject(&proc{|a,b| [a,b].max_by{|m| m.length}}) == ['to','four','tri'].inject(&proc{|a,b| [a,b].max_by{|m| m.length}})`
- [ ] #my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block
      `[1,2,3].inject(&proc{|total, num| total*num}) == [1,2,3].my_inject(&proc{|total, num| total*num})`
- [ ] #my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block::range
      `(1..3).inject(&proc{|total, num| total*num}) == (1..3).my_inject(&proc{|total, num| total*num})`
- [ ] #my_inject when a block is given with an initail value combines all elements of enum by applying a binary operation, specified by a 
block
      `(1..3).inject(4) { |prod, n| prod * n } == (1..3).my_inject(4) { |prod, n| prod * n }`
- [ ] #my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method:range
      `(1..9).inject(:+) == (1..9).my_inject(:+)`
- [ ] #my_inject when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method::range
      `(1..3).inject(4, :*) == (1..3).my_inject(4, :*)`
- [ ] #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject
    `multiply_els([1,2,3]) == 6`
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

**Comment by RyelBanfield** [comment]
image

I think we need another review as the very first correction we are told to make doesn't make any sense. my_each is supposed to work exactly like the original .each which it does as shown in the photo attached. Maybe I'm misunderstanding.

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Mert Gündüz & Ryel Banfield** 
Nice job with the improvements to your project :muscle:  :muscle:
My bad, I did a mistake in the feedback for that particular error
Still, the feedback for the other errors was ok and you didn't make the requested improvements. For example, "multiply_els([1,2,3]) == 6", this one was failing before and is still failing now. Please, don't stop if you find any other error in the feedback, check all the points before requesting another review:

- [x] #my_each when a block is given when self is a range returns the range itself after calling the given block once for each element in 
self
        `Range.new(1, 5).my_each(&proc{|x| x>2}) == Range.new(1, 5).each(&proc{|x| x>2})`
- [x] #my_all? returns false if the block returns false or nil::range
    `(1..3).all?(&proc{|x| x%2==0}) == (1..3).my_all?(&proc{|x| x%2==0})`
- [x] #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern      
      `[5,5,[5]].all?(5) == [5,5,[5]].my_all?(5)`
- [ ] #my_any? returns false if the block ever returns a value false or nil::range
    `(1..3).any?(&proc{|x| x==0}) == (1..3).my_any?(&proc{|x| x==0})`
- [ ] #my_any? when no block or argument is given returns false if at least one of the collection is not true
      `[false, 0].any? == [false, 0].my_any?`
- [ ] #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern
      `['cat', 'dog','car'].any?('cat') == ['cat', 'dog','car'].my_any?('cat')`
- [ ] #my_none? returns false if the block ever returns true for all elements::range
    `(1..3).none?(&proc{|num| num%2==0}) == (1..3).my_none?(&proc{|num| num%2==0})`
- [ ] #my_none? when a Regex is passed as an argument returns false if any of the collection
matches the Regex
      `['dog', 'cat'].none?(/d/) == ['dog', 'cat'].my_none?(/d/)`
- [ ] #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern       
      `[5,'dog','car'].none?(5) == [5,'dog','car'].my_none?(5)`
- [ ] #my_count returns the number of items in enum through enumeration::range
    `(1..3).count == (1..3).my_count`
- [ ] #my_map returns a new array with the results of running a given block once for every element in enum::range
    `(1..3).map(&proc{|num| num+1}) == (1..3).my_map(&proc{|num| num+1})`
- [ ] #my_inject searches for the longest word in an array of strings
ength}})`
- [ ] #my_inject when a block is given without an initail value combines all elements of enum      `[1,2,3].inject(&proc{|total, num| total*num}) == [1,2,3].my_inject(&proc{|total, num| total*num})`
- [ ] #my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block::range
      `(1..3).inject(&proc{|total, num| total*num}) == (1..3).my_inject(&proc{|total, num| total*num})`
- [ ] #my_inject when a block is given with an initail value combines all elements of enum by applying a binary operation, specified by a 
block
      `(1..3).inject(4) { |prod, n| prod * n } == (1..3).my_inject(4) { |prod, n| prod * n }`
- [ ] #my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method:range
      `(1..9).inject(:+) == (1..9).my_inject(:+)`
- [ ] #my_inject when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method::range
      `(1..3).inject(4, :*) == (1..3).my_inject(4, :*)`
- [ ] #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject
    `multiply_els([1,2,3]) == 6`
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    false
+  end
+
+  def my_none?(param = nil)
+    if block_given?
+      !my_any?(param)
+    else
+      !my_any?(param)
+    end
+  end
```
Comment (enaburekhan) [inline enumerables.rb:66]: #my_none? returns true if the block never returns true for all elements
expected: true
            got: false
Refactor the highlighted code to account for the above test.

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Required changes:full_moon:
**Hi team!
Great job so far:clap::clap:.  The way you reuse methods in the codebase shows your excellent mastery of this milestone. Keep it up:thumbsup:. Just one test is failing. Let's fix it and you can move on with your next task:blush:.**

- [ ] #my_none? returns true if the block never returns true for all elements. It's currently returning false instead of true.

As already highlighted in the PR below. Refactor the highlighted code to account for the test.

Make the above-required change and submit it for another review.
Kindly drop your questions :point_down: and ping me on slack @ Eric Enaburekhan to respond.
Happy Coding:blush:
Eric Enaburekhan

**Review by umairarshadbutt — CHANGES_REQUESTED** [review]
# Status: REQUIRED CHANGES :warning: :construction: 

Hi 
Good job with the changes! :tada: We are almost there.
We, however, do have a few changes to carry out before we proceed (see review comment).

##

 For this project, we have to make sure nothing is messed up. Kindly compare the screenshots below 👇 and work on the suggested changes.
- [ ] `#my_none?` returns true if the block never returns true for all elements
- [ ] `#my_none?` when no block or argument is given returns true only if none of the collection members is true
- [ ] `#my_none?` when a Regex is passed as an argument returns true if none of the collection matches the Regex
- [ ] `#my_none?` when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
##
- **When you have made the specified changes, please submit another code review. Thank you!**
##
**If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

**Comment by mgunduz1** [comment]
> # Status: REQUIRED CHANGES  
> Hi
> Good job with the changes!  We are almost there.
> We, however, do have a few changes to carry out before we proceed (see review comment).
> 
> For this project, we have to make sure nothing is messed up. Kindly compare the screenshots below  and work on the suggested changes.
> 
> * [ ]  `#my_none?` returns true if the block never returns true for all elements
> * [ ]  `#my_none?` when no block or argument is given returns true only if none of the collection members is true
> * [ ]  `#my_none?` when a Regex is passed as an argument returns true if none of the collection matches the Regex
> * [ ]  `#my_none?` when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
> 
> * **When you have made the specified changes, please submit another code review. Thank you!**
> 
> **If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
> HAPPY CODING 
> **[Umair Arshad](https://umairarshad.me/)**
> [Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

There is no screenshot in your comment. With which screenshots exactly you want us to compare with our work ?

**Review by arslanbisharat — APPROVED** [review]
### Approved 👏 

Thank you for making the requested changes. You have added all methods. Please move forward with your next project.

**Best,
Arslan Bisharat**

**Pull request merged**